cachectl/activepages.go

```diff
@@ -1,74 +1,55 @@
 package cachectl
 
-/*
-#include <stdio.h>
-#include <stdlib.h>
-#include <unistd.h>
-#include <fcntl.h>
-#include <sys/types.h>
-#include <sys/stat.h>
-#include <sys/mman.h>
-
-int activePages(const char *path)
-{
-    int i, j, fd, pages, pagesize;
-    struct stat st;
-    void *m;
-    char *pageinfo;
-
-    fd = open(path, O_RDONLY);
-    if(fd == -1) {
-        return -1;
-    }
-
-    if(fstat(fd, &st) == -1) {
-        goto error;
-    }
-
-    pagesize = getpagesize();
-    pages = (st.st_size + pagesize - 1) / pagesize;
-    pageinfo = calloc(sizeof(*pageinfo), pages);
-    if(!pageinfo) {
-        goto error;
-    }
-
-    m = mmap(NULL, st.st_size, PROT_NONE, MAP_SHARED, fd, 0);
-    if(m == MAP_FAILED) {
-        free(pageinfo);
-        goto error;
-    }
-
-    if(mincore(m, st.st_size, pageinfo) == -1) {
-        free(pageinfo);
-        munmap(m, st.st_size);
-        goto error;
-    }
-
-    i = 0;
-    j = 0;
-    for (i = 0; i < pages; i++) {
-        if(pageinfo[i] & 1) {
-            j++;
-        }
-    }
-
-    free(pageinfo);
-    munmap(m, st.st_size);
-
-    return j;
-error:
-    close(fd);
-    return -1;
-}
-*/
-import "C"
 import (
+	"os"
 	"unsafe"
+
+	"golang.org/x/sys/unix"
 )
 
-func activePages(path string) int {
-	cs := C.CString(path)
-	result := C.activePages(cs)
-	C.free(unsafe.Pointer(cs))
-	return int(result)
+func activePages(path string) (int, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return 0, err
+	}
+	fsize := fi.Size()
+
+	if fsize == 0 {
+		return 0, nil
+	}
+
+	mmap, err := unix.Mmap(int(f.Fd()), 0, int(fsize), unix.PROT_NONE, unix.MAP_SHARED)
+	if err != nil {
+		return 0, err
+	}
+
+	pagesize := int64(os.Getpagesize())
+	pages := (fsize + pagesize - 1) / pagesize
+	pageinfo := make([]byte, pages)
+
+	mmapPtr := uintptr(unsafe.Pointer(&mmap[0]))
+	sizePtr := uintptr(fsize)
+	pageinfoPtr := uintptr(unsafe.Pointer(&pageinfo[0]))
+
+	ret, _, err := unix.Syscall(unix.SYS_MINCORE, mmapPtr, sizePtr, pageinfoPtr)
+	if ret != 0 {
+		return 0, err
+	}
+	defer unix.Munmap(mmap)
+
+	result := 0
+
+	for _, p := range pageinfo {
+		if p%2 == 1 {
```

Comment:
Original code looks easier to understand. Because each entry of pageinfo is assumed that it is checked by the least significant bit. So using bit manipulation is more clear.

According to **man mincore**, 

```
On return, the least significant bit of each byte will be set if the corresponding page is currently resident in memory, and be clear otherwise.  (The settings of the other bits in each byte are undefined; these bits are reserved for possible  later  use.)
```